Feature/bt 154 treasury reward

mzxyz opened the pull request:
### Award all transaction fee equal to all validators

#### Implementation:

* Rewards are `SplitTwoWays` by 0 part to the treasury and 1 part to the validators 
* Reward is then divided by the currently elected validators
* Each validator are rewarded their share with `deposit_creating`

#### Test cases:

*  Add gensis config for `staking` and  `session` modules in `mock.rs`
*  Test case: `staking_genesis_config_works`
*  Test case : `validators_received_equal_tx_fee_reward`

roy-powerfinance committed:
Partial work for rewarding all transaction fees to all validators

roy-powerfinance committed:
Partial work for rewarding all transaction fees to all validators

Commit:
Fix build issues. Fix formatting.

Commit:
Fix build issues. Fix formatting.

mzxyz committed:
genesis config for staking in runtime test mock

roy-powerfinance committed:
Moved the validator struct from libs to impl
Added logic to accumulate positive imbalance so rewards distributed to
the validators are tracked.

roy-powerfinance committed:
Merge remote-tracking branch 'origin/develop' into prototype/bt_154_treasury_reward

* origin/develop:
  Better Transaction Payment Errors (#142)
  Add setup for integration tests and 12 tests (#140)
  update return error for buy_fee_asset (#141)
  Add docker link (#136)
  Update to the changes in BuyFeeAsset (#138)

# Conflicts:
#	runtime/tests/tests.rs
#	testing/src/keyring.rs

roy-powerfinance committed:
Merge remote-tracking branch 'origin/prototype/bt_154_treasury_reward' into feature/bt_154_treasury_reward

* origin/prototype/bt_154_treasury_reward:
  Moved the validator struct from libs to impl Added logic to accumulate positive imbalance so rewards distributed to the validators are tracked.
  Fix build issues. Fix formatting.
  Partial work for rewarding all transaction fees to all validators

# Conflicts:
#	runtime/src/lib.rs

mzxyz committed:
add staking test cases

roy-powerfinance committed:
Renamed the Validators struct to SplitToAllValidators
Added some comments for the code
Removed the balance module reference from the runtime module

roy-powerfinance committed:
Removed duplicate define for NegativeImbalance
Added logic where if the negative imbalance is zero, return
ran cargo fmt

roy-powerfinance committed:
Merge remote-tracking branch 'origin/feature/zh/bt_154_treasury_reward' into feature/bt_154_treasury_reward

* origin/feature/zh/bt_154_treasury_reward:
  add staking test cases
  genesis config for staking in runtime test mock

Commit:
update validators fee reward test cases (#149)

royyang-centrality commented on runtime/tests/tests.rs:
```diff
@@ -120,6 +121,74 @@ fn transfer_fee<E: Encode>(extrinsic: &E, fee_multiplier: Fixed64, runtime_call:
+}
+
+#[test]
+fn staking_validators_should_received_equal_transaction_fee_reward() {
```
@ @

cowboy-bebug commented on runtime/Cargo.toml:
```diff
@@ -73,7 +73,9 @@ wasm-builder-runner = { package = "substrate-wasm-builder-runner", version = "1.
 cennznet-testing = { path = "../testing"}
+cennznet-cli = { path = "../cli" }
 sp-io = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "1.0.0-rc2" }
+crml-sylo = { path = "../crml/sylo", default-features = false }
```
I think this should be removed

mzxyz committed:
update test case: total_issuance is ajuested
[ci skip]

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,45 @@ use sp_runtime::{
+impl OnUnbalanced<NegativeImbalance> for SplitToAllValidators {
+	fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+		let validators = Staking::<Runtime>::current_elected();
+		if validators.len() > 0 || amount.peek().is_zero(){
```
should this be
```
if validators.len().is_zero() || amount.peek().is_zero() {
    return;
}
```

cowboy-bebug replied:
I thought so as well

roy-powerfinance replied:
yep fixed

jordy25519 commented on runtime/tests/mock.rs:
```diff
@@ -15,30 +15,46 @@
+	]
+}
+
+// get all validators (slash account , controller account)
```
'stash account'

jordy25519 commented on runtime/tests/tests.rs:
```diff
@@ -120,6 +121,85 @@ fn transfer_fee<E: Encode>(extrinsic: &E, fee_multiplier: Fixed64, runtime_call:
+			assert!(r.is_ok());
+
+			// Check total_issurance is adjusted
+			// FIXME: total_issurance is not adjusted
```
Since we mint exactly what was burned (except remainder) total issuance won't change until we introduce an inflation parameter in the reward step

roy-powerfinance replied:
Reward paid out now creates a positive imbalance that offsets the negative imbalance, so if there are remainders left, the overall issuance is adjusted.

hoani commented on runtime/tests/tests.rs:
```diff
@@ -120,6 +121,74 @@ fn transfer_fee<E: Encode>(extrinsic: &E, fee_multiplier: Fixed64, runtime_call:
+		.execute_with(|| {
+			for validator in validators() {
+				let (stash, controller) = validator;
+				let staked_amount = balance_amount / 5;
```
Please comment that this value comes from Mock (ln 119)

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,45 @@ use sp_runtime::{
+			let asset_id = amount.asset_id();
+			for v in &validators {
+				let payout = <GenericAsset<Runtime> as MultiCurrencyAccounting>::deposit_creating(
+					&v,
```
This should pay to the validator reward destination account: https://github.com/plugblockchain/plug-blockchain/blob/d2319b9c98c209be11544e5b05b4797e670c5be6/frame/staking/src/lib.rs#L696

roy-powerfinance replied:
Implemented the required logic

cowboy-bebug commented on runtime/tests/mock.rs:
```diff
@@ -82,13 +100,48 @@ impl ExtBuilder {
 		.assimilate_storage(&mut t)
 		.unwrap();
+
+		let stash = self.initial_balance / 5;
```
Instead of this pattern, we could add `stash` field in `struct ExtBuilder` and a setter method (up to u)

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -120,6 +121,85 @@ fn transfer_fee<E: Encode>(extrinsic: &E, fee_multiplier: Fixed64, runtime_call:
+fn staking_genesis_config_works() {
+	let balance_amount = 10_000 * TransactionBaseFee::get();
+	ExtBuilder::default()
+		.initial_balance(balance_amount)
```
We could add `.stash(balance_amount / 5)` here if we use the ExtBuilder setter pattern

jordy25519 reviewed (commented):
Looks good so far!
Could add some unit tests in impls.rs for the 0 reward case
Also would want to respect the validators preferred payment account

aliXsed commented on cli/src/chain_spec/mod.rs:
```diff
@@ -46,19 +46,22 @@ pub mod rimu;
 type AccountPublic = <Signature as Verify>::Signer;
 
+/// A type contains authority keys
+pub type AuthorityKeys = (
```
Can you add a short line doc for every field of this tuple?

mzxyz committed:
set stash amount as configuration
[ci skip]

mzxyz committed:
fix build issue

cowboy-bebug commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,45 @@ use sp_runtime::{
+>>::NegativeImbalance;
+pub type PositiveImbalance = <<Runtime as crml_transaction_payment::Trait>::Currency as Currency<
+	<Runtime as frame_system::Trait>::AccountId,
+>>::PositiveImbalance;
```
This can be removed if we define them in `runtime/src/lib.rs` like this:
```rust
type PositiveImbalance = <GenericAsset as MultiCurrencyAccounting>::PositiveImbalance;
type NegativeImbalance = <GenericAsset as MultiCurrencyAccounting>::NegativeImbalance;
```
and import the above using `use crate::{PositiveImbalance, NegativeImbalance};`

roy-powerfinance replied:
done

roy-powerfinance committed:
Fixed the bug that the positive imbalance does not change the overall issuance
Made some logic change to the validator reward
Moved the megative and positive imbalance to libs

roy-powerfinance committed:
Added logic that gets the correct RewardDestination when paying validators

cowboy-bebug commented on runtime/src/impls.rs:
```diff
@@ -47,26 +48,32 @@ pub struct SplitToAllValidators;
+		for validator in &validators {
+			let dest = Staking::<Runtime>::payee(validator);
+			let reward_destination_account_id = match dest {
+				RewardDestination::Controller => Staking::<Runtime>::bonded(validator).unwrap(),
```
Maybe we want to use
```rust
Staking::<Runtime>::bonded(validator).unwrap_or_else(|| validator);
```
to ensure that if `bonded() -> None` then we just want to use validator

roy-powerfinance replied:
Yep done.

roy-powerfinance committed:
Merged from develop to resolve conflicts.

roy-powerfinance committed:
Changed an unwrap to unwrap_or_else to handle the None case

aliXsed commented on runtime/src/impls.rs:
```diff
@@ -17,24 +17,66 @@
+/// The remainder from the division are burned.
+impl OnUnbalanced<NegativeImbalance> for SplitToAllValidators {
+	fn on_nonzero_unbalanced(imbalance: NegativeImbalance) {
+		let validators = Staking::<Runtime>::current_elected();
```
Can we be certain that on_nonzero_unbalanced would be called before a new set of validators are elected for a new era? In other words, are we rewarding the correct set of validators? [It's just a question]

roy-powerfinance replied:
this function is only called during validate(...) function, which I think is used to verify transactions. Therefore it will only be called when we are processing transactions - i.e. during an session when validators are all set and present.
Worse case scenario we have no validators and we just burn the transaction fee

roy-powerfinance replied:
Also since the chain is running single threaded, I think it would not be possible for the validators to change in the middle of a transaction.
In addition, we make a copy of the list of validators, therefore even if the validators change, we would be OK

aliXsed commented on runtime/src/impls.rs:
```diff
@@ -17,24 +17,66 @@
+			return;
+		}
+		// Get a list of elected validators
+		let per_validator_reward: Balance = imbalance.peek() / (validators.len() as Balance);
```
```len() as Balance``` is a bad hack IMO. "Balance" should be divisible by a number. If it's not true we should address the root cause.

roy-powerfinance replied:
Discussed with Alex, and we cannot find a better way of doing this. Casting it into Balance means our datatype remains flexible. We will keep as is

cowboy-bebug committed:
Parameterise validator_count for mock

jordy25519 commented on runtime/tests/mock.rs:
```diff
@@ -15,30 +15,65 @@
+	for i in 0..n as usize {
+		authority_keys.push(get_authority_keys_from_seed(accounts[i]))
+	}
+	authority_keys
```
```suggestion
let accounts = vec!["Alice", "Bob", "Charlie", "Dave", "Eve", "Ferdie"];
accounts[..n].map(|s| get_authority_keys_from_seed(s)).collect()

```

cowboy-bebug committed:
Update runtime/tests/mock.rs

Co-Authored-By: Jordan <beauchjord@gmail.com>

cowboy-bebug committed:
Format

roy-powerfinance commented on runtime/tests/mock.rs:
```diff
@@ -27,30 +27,45 @@ use sp_runtime::Perbill;
-		get_authority_keys_from_seed("Bob"),
-		get_authority_keys_from_seed("Charlie"),
-	]
+pub fn generate_initial_authorities(n: usize) -> Vec<AuthorityKeys> {
```
Since this is a public function, we cannot ensure that n <= 6. I think we should do some kind of boundary checks

cowboy-bebug replied:
Thanks for spotting this!

cowboy-bebug committed:
Add bound check and fix a bug

jordy25519 reviewed (approved):
LGTM
